File: feedstamer/flutter_app/lib/models/user_model.dart
import 'package:cloud_firestore/cloud_firestore.dart';

class UserModel {
  final String uid;
  final String email;
  final String name;
  final String profilePicture;
  final SubscriptionModel subscription;
  final PreferencesModel preferences;
  final DateTime createdAt;
  
  UserModel({
    required this.uid,
    required this.email,
    required this.name,
    required this.profilePicture,
    required this.subscription,
    required this.preferences,
    required this.createdAt,
  });
  
  factory UserModel.fromMap(Map<String, dynamic> map) {
    return UserModel(
      uid: map['uid'] ?? '',
      email: map['email'] ?? '',
      name: map['name'] ?? '',
      profilePicture: map['profilePicture'] ?? '',
      subscription: SubscriptionModel.fromMap(map['subscription'] ?? {}),
      preferences: PreferencesModel.fromMap(map['preferences'] ?? {}),
      createdAt: (map['createdAt'] as Timestamp?)?.toDate() ?? DateTime.now(),
    );
  }
  
  Map<String, dynamic> toMap() {
    return {
      'uid': uid,
      'email': email,
      'name': name,
      'profilePicture': profilePicture,
      'subscription': subscription.toMap(),
      'preferences': preferences.toMap(),
      'createdAt': Timestamp.fromDate(createdAt),
    };
  }
  
  UserModel copyWith({
    String? uid,
    String? email,
    String? name,
    String? profilePicture,
    SubscriptionModel? subscription,
    PreferencesModel? preferences,
    DateTime? createdAt,
  }) {
    return UserModel(
      uid: uid ?? this.uid,
      email: email ?? this.email,
      name: name ?? this.name,
      profilePicture: profilePicture ?? this.profilePicture,
      subscription: subscription ?? this.subscription,
      preferences: preferences ?? this.preferences,
      createdAt: createdAt ?? this.createdAt,
    );
  }
}

class SubscriptionModel {
  final String type; // 'free', 'premium', 'family', 'enterprise'
  final String status; // 'active', 'cancelled', 'expired'
  final DateTime? startDate;
  final DateTime? endDate;
  
  SubscriptionModel({
    required this.type,
    required this.status,
    this.startDate,
    this.endDate,
  });
  
  factory SubscriptionModel.fromMap(Map<String, dynamic> map) {
    return SubscriptionModel(
      type: map['type'] ?? 'free',
      status: map['status'] ?? 'active',
      startDate: (map['startDate'] as Timestamp?)?.toDate(),
      endDate: (map['endDate'] as Timestamp?)?.toDate(),
    );
  }
  
  Map<String, dynamic> toMap() {
    return {
      'type': type,
      'status': status,
      'startDate': startDate != null ? Timestamp.fromDate(startDate!) : null,
      'endDate': endDate != null ? Timestamp.fromDate(endDate!) : null,
    };
  }
  
  bool get isPremium => type != 'free';
  
  bool get isActive => status == 'active';
}

class PreferencesModel {
  final String theme; // 'light', 'dark', 'system'
  final NotificationsPreferencesModel notifications;
  final FeedsSettingsModel FeedsSettings;
  final SessionLimitsModel sessionLimits;
  
  PreferencesModel({
    required this.theme,
    required this.notifications,
    required this.FeedsSettings,
    required this.sessionLimits,
  });
  
  factory PreferencesModel.fromMap(Map<String, dynamic> map) {
    return PreferencesModel(
      theme: map['theme'] ?? 'system',
      notifications: NotificationsPreferencesModel.fromMap(map['notifications'] ?? {}),
      FeedsSettings: FeedsSettingsModel.fromMap(map['FeedsSettings'] ?? {}),
      sessionLimits: SessionLimitsModel.fromMap(map['sessionLimits'] ?? {}),
    );
  }
  
  Map<String, dynamic> toMap() {
    return {
      'theme': theme,
      'notifications': notifications.toMap(),
      'FeedsSettings': FeedsSettings.toMap(),
      'sessionLimits': sessionLimits.toMap(),
    };
  }
  
  PreferencesModel copyWith({
    String? theme,
    NotificationsPreferencesModel? notifications,
    FeedsSettingsModel? FeedsSettings,
    SessionLimitsModel? sessionLimits,
  }) {
    return PreferencesModel(
      theme: theme ?? this.theme,
      notifications: notifications ?? this.notifications,
      FeedsSettings: FeedsSettings ?? this.FeedsSettings,
      sessionLimits: sessionLimits ?? this.sessionLimits,
    );
  }
}

class NotificationsPreferencesModel {
  final bool enabled;
  final Map<String, bool> types;
  
  NotificationsPreferencesModel({
    required this.enabled,
    required this.types,
  });
  
  factory NotificationsPreferencesModel.fromMap(Map<String, dynamic> map) {
    final typesMap = map['types'];
    Map<String, bool> typesData = {};
    
    if (typesMap is Map) {
      typesMap.forEach((key, value) {
        if (value is bool) {
          typesData[key] = value;
        }
      });
    }
    
    return NotificationsPreferencesModel(
      enabled: map['enabled'] ?? true,
      types: typesData.isEmpty 
        ? {
            'newContent': true,
            'accountActivity': true,
            'usageReminders': true,
          }
        : typesData,
    );
  }
  
  Map<String, dynamic> toMap() {
    return {
      'enabled': enabled,
      'types': types,
    };
  }
}

class FeedsSettingsModel {
  final String defaultView; // 'unified', 'platform'
  final String contentOrder; // 'chronological', 'platform'
  final bool showReadPosts;
  
  FeedsSettingsModel({
    required this.defaultView,
    required this.contentOrder,
    required this.showReadPosts,
  });
  
  factory FeedsSettingsModel.fromMap(Map<String, dynamic> map) {
    return FeedsSettingsModel(
      defaultView: map['defaultView'] ?? 'unified',
      contentOrder: map['contentOrder'] ?? 'chronological',
      showReadPosts: map['showReadPosts'] ?? false,
    );
  }
  
  Map<String, dynamic> toMap() {
    return {
      'defaultView': defaultView,
      'contentOrder': contentOrder,
      'showReadPosts': showReadPosts,
    };
  }
}

class SessionLimitsModel {
  final bool enabled;
  final int dailyLimit; // minutes
  final int reminderInterval; // minutes
  
  SessionLimitsModel({
    required this.enabled,
    required this.dailyLimit,
    required this.reminderInterval,
  });
  
  factory SessionLimitsModel.fromMap(Map<String, dynamic> map) {
    return SessionLimitsModel(
      enabled: map['enabled'] ?? false,
      dailyLimit: map['dailyLimit'] ?? 60,
      reminderInterval: map['reminderInterval'] ?? 20,
    );
  }
  
  Map<String, dynamic> toMap() {
    return {
      'enabled': enabled,
      'dailyLimit': dailyLimit,
      'reminderInterval': reminderInterval,
    };
  }
}

File: feedstamer/backend/src/models/user.model.js
const mongoose = require('mongoose');
const bcrypt = require('bcryptjs');
const validator = require('validator');

const userSchema = new mongoose.Schema({
  email: {
    type: String,
    required: [true, 'Email is required'],
    unique: true,
    lowercase: true,
    validate: [validator.isEmail, 'Please provide a valid email']
  },
  password: {
    type: String,
    required: [true, 'Password is required'],
    minlength: 8,
    select: false
  },
  name: {
    type: String,
    required: [true, 'Name is required']
  },
  profilePicture: {
    type: String,
    default: ''
  },
  subscription: {
    type: {
      type: String,
      enum: ['free', 'premium', 'family', 'enterprise'],
      default: 'free'
    },
    startDate: {
      type: Date
    },
    endDate: {
      type: Date
    },
    status: {
      type: String,
      enum: ['active', 'cancelled', 'expired'],
      default: 'active'
    }
  },
  connectedPlatforms: [{
    platform: {
      type: String,
      enum: ['twitter', 'instagram', 'youtube', 'reddit', 'linkedin']
    },
    accessToken: {
      type: String
    },
    refreshToken: {
      type: String
    },
    tokenExpiry: {
      type: Date
    },
    userId: {
      type: String
    },
    username: {
      type: String
    },
    connected: {
      type: Boolean,
      default: true
    }
  }],
  preferences: {
    theme: {
      type: String,
      enum: ['light', 'dark', 'system'],
      default: 'system'
    },
    notifications: {
      enabled: {
        type: Boolean,
        default: true
      },
      types: {
        newContent: {
          type: Boolean,
          default: true
        },
        accountActivity: {
          type: Boolean,
          default: true
        },
        usageReminders: {
          type: Boolean,
          default: true
        }
      }
    },
    FeedsSettings: {
      defaultView: {
        type: String,
        enum: ['unified', 'platform'],
        default: 'unified'
      },
      contentOrder: {
        type: String,
        enum: ['chronological', 'platform'],
        default: 'chronological'
      },
      showReadPosts: {
        type: Boolean,
        default: false
      }
    },
    sessionLimits: {
      enabled: {
        type: Boolean,
        default: false
      },
      dailyLimit: {
        type: Number,
        default: 60 // minutes
      },
      reminderInterval: {
        type: Number,
        default: 20 // minutes
      }
    }
  },
  analytics: {
    timeSpent: [{
      date: {
        type: Date
      },
      minutes: {
        type: Number,
        default: 0
      },
      platform: {
        type: String
      }
    }],
    contentViewed: [{
      date: {
        type: Date
      },
      count: {
        type: Number,
        default: 0
      },
      platform: {
        type: String
      }
    }]
  },
  createdAt: {
    type: Date,
    default: Date.now
  },
  passwordChangedAt: Date,
  passwordResetToken: String,
  passwordResetExpires: Date,
  active: {
    type: Boolean,
    default: true,
    select: false
  }
});

// Hash the password before saving
userSchema.pre('save', async function(next) {
  // Only run this function if password was modified
  if (!this.isModified('password')) return next();

  // Hash the password with cost of 12
  this.password = await bcrypt.hash(this.password, 12);

  next();
});

userSchema.pre('save', function(next) {
  if (!this.isModified('password') || this.isNew) return next();

  this.passwordChangedAt = Date.now() - 1000;
  next();
});

userSchema.pre(/^find/, function(next) {
  // this points to the current query
  this.find({ active: { $ne: false } });
  next();
});

// Instance method to check if password is correct
userSchema.methods.correctPassword = async function(candidatePassword, userPassword) {
  return await bcrypt.compare(candidatePassword, userPassword);
};

// Check if password was changed after token was issued
userSchema.methods.changedPasswordAfter = function(JWTTimestamp) {
  if (this.passwordChangedAt) {
    const changedTimestamp = parseInt(
      this.passwordChangedAt.getTime() / 1000,
      10
    );
    return JWTTimestamp < changedTimestamp;
  }
  return false;
};

const User = mongoose.model('User', userSchema);

module.exports = User;